app/coffee/RoomManager.coffee

```diff
@@ -0,0 +1,93 @@
+logger = require 'logger-sharelatex'
+metrics = require "metrics-sharelatex"
+{EventEmitter} = require 'events'
+
+IdMap = new Map() # keep track of whether ids are from projects or docs
+RoomEvents = new EventEmitter() # emits {project,doc}-active and {project,doc}-empty events
+
+# Manage socket.io rooms for individual projects and docs
+#
+# The first time someone joins a project or doc we emit a 'project-active' or
+# 'doc-active' event.
+#
+# When the last person leaves a project or doc, we emit 'project-empty' or
+# 'doc-empty' event.
+#
+# The pubsub side is handled by ChannelManager
+
+module.exports = RoomManager =
+
+    joinProject: (client, project_id, callback = () ->) ->
+        @joinEntity client, "project", project_id, callback
+
+    joinDoc: (client, doc_id, callback = () ->) ->
+        @joinEntity client, "doc", doc_id, callback
+
+    leaveDoc: (client, doc_id) ->
+        @leaveEntity client, "doc", doc_id
+
+    leaveProjectAndDocs: (client) ->
+        # what rooms is this client in? we need to leave them all. socket.io
+        # will cause us to leave the rooms, so we only need to manage our
+        # channel subscriptions... but it will be safer if we leave them
+        # explicitly, and then socket.io will just regard this as a client that
+        # has not joined any rooms and do a final disconnection.
+        for id in @_roomsClientIsIn(client)
+            entity = IdMap.get(id)
+            @leaveEntity client, entity, id
+
+    emitOnCompletion: (promiseList, eventName) ->
+        result = Promise.all(promiseList)
+        result.then () -> RoomEvents.emit(eventName)
+        result.catch (err) -> RoomEvents.emit(eventName, err)
+
+    eventSource: () ->
+        return RoomEvents
+
+    joinEntity: (client, entity, id, callback) ->
+        beforeCount = @_clientsInRoom(client, id)
+        # is this a new room? if so, subscribe
+        if beforeCount == 0
+            logger.log {entity, id}, "room is now active"
+            RoomEvents.once "#{entity}-subscribed-#{id}", (err) ->
+                # only allow the client to join when all the relevant channels have subscribed
```

Comment:
If there are ops processed between the socket.io connection and the joining of the room does the client need to reconnect?

Reply:
Hmm. Good point. 

I changed the ordering at [WebsocketController.coffee#L95-L99](https://github.com/overleaf/real-time/blob/4c81e407d2fd3bf158dfb310c82a6268d3cf1c27/app/coffee/WebsocketController.coffee#L95-L99) so that the channel is subscribed and client goes into the room before we request the doc from docupdater and send it to them - this ensures that they don't miss any ops that have been processed after the doc is requested from docupdater.   But it does mean the client could be sent an applied-op before they have the doc and I had overlooked that.  I have just checked in web and this is handled ok, the web frontend to will discard it.

We bind to the socket events early  at [Document.js#L102](https://github.com/overleaf/web-internal/blob/b52fcdd551e6fe0a75a4b051237a9d5f483deaf9/public/src/ide/editor/Document.js#L102), before even joining the doc.  Updates arriving before the doc is joined are handled because the front end only accept updates when `doc` is not null at  [Document.js#L422](https://github.com/overleaf/web-internal/blob/b52fcdd551e6fe0a75a4b051237a9d5f483deaf9/public/src/ide/editor/Document.js#L422), so nothing gets applied until the doc is fully  joined.  So we are ok on this one.